File: 2 FORTRAN/Task_1/main_Jun01_Task_1.f90
! Jun 9, 2017 in IVF & VS
! June 1,2017
! May 22-24, 2017
! 20170411 PM
! Mar18,2017
! Task #20170221
! Related to May16/Oct17,2016
! edited in Feb21,2017
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
module point_para
    ! pointer to get parameters
    real(kind=8),pointer :: p_para(:)
    character(len=:), pointer :: p_typen
    real(kind=8),pointer :: p_epu
end module
    
program main_Jun01_Task_1
    ! Include the IMSL lib
    include 'link_fnl_shared.h'
    ! use module
    use point_para

    implicit none
    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    ! declare paramaters
    character(len=7),parameter :: taskcode = "Jun01T1"
    character(len=3),parameter :: num_para = "V27"
    real(kind=4),parameter :: pi = 3.1415926
    !!!!!!!!!!!!!!!!!!!
    ! declare function
    character(len=10), external :: cell_type
    character(len=10), external :: cell_case
    
    real,intrinsic :: sqrt
    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    ! declare variable
    integer I
    ! for the loop
    
    character(len=3), target :: typen
    character(len=10):: casecode
    character(len=:),allocatable :: str_s
    character(len=:),allocatable :: str_fig
    character(len=13) :: str_para 
    
    real(kind=8),target :: epsilon_u
    
    real(kind=8),TARGET :: parameters(15)
    ! parameters=[A;B;r;c_p;beta;phi;delta;sigma;lambda;pstar;b;c_f;mu;B1;B2];
    
    real(kind=4),allocatable :: vt(:)
    real(kind=8),allocatable :: vepsilon_d(:)
    real(kind=8),allocatable :: vepsilon_c(:)
    real(kind=8),allocatable :: vtheta(:)
    real(kind=8),allocatable :: valpha(:)
    logical(kind=4),allocatable :: vexitflag(:)
    
    real(kind=8),allocatable :: vu(:)
    real(kind=8),allocatable :: vhs(:)
    real(kind=8),allocatable :: vhn(:)
    
    real(kind=8) :: pstar
    integer tt,cc
    ! tt is # of type of function F(x)
    ! cc is # of case
    
    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    p_para=>parameters
    p_typen=>typen
    P_epu=>epsilon_u
    
    str_para = 'paraJun01_'//num_para
    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    tt=3
    ! type of F(x)
    loop2: do cc= 8,10
    ! case #
        !!!!!!!!!!!!!!!!!!!!!!!!!!!!
        typen = cell_type(tt)
        casecode = cell_case(cc)
    
        if (typen=="I") then
            epsilon_u = sqrt(3.0)
        else if (typen=="II") then
            epsilon_u=sqrt(2.0)/(sqrt(pi-2.0))
        else if (typen=="III") then
            epsilon_u=1.0
        else if (typen=="IV") then
            epsilon_u=(3.0-sqrt(3.0))/2.0
        else if (typen=="O") then
            epsilon_u=(pi/4.0)/sin(pi/4.0)
        end if
    
        ! get the parameters
        call MyParameter_Jun01V27(parameters)
        !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        ! check the pointers and targets
        write(*,"('Parameters are:',15(/f8.3))") parameters
        write(*,"('Parameters are:',15(/f8.3))") P_PARA
        !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
        ! change parameters with different cases
        if (casecode=="pstar") then ! 1
            vt = (/ (0.01*i,i = 18,250) /) 
            str_s="pstar"
        else if (casecode=="b") then ! 2
            vt = (/ (0.01*i,i = 1,100) /)
            str_s = "b"
        else if (casecode=="phi") then ! 3
            vt = (/ (0.01*i,i = 30,95) /)
            str_s = "phi"
        else if (casecode=="sigma") then ! 4
            vt = (/ (0.01*i,i = 1,120) /)
            str_s = "sigma"
        else if (casecode=="beta") then ! 5
            vt = (/ (0.01*i,i = 40,99) /)
            str_s = "beta"
        else if (casecode=="lambda") then ! 6
            vt = (/ (0.01*i,i = 2,63) /)
            str_s = "lambda"
        else if (casecode=="c_f") then ! 7
            vt = (/ (0.01*i,i = 12,130) /)
            str_s = "c_F"
        else if (casecode=="r") then ! 8
            vt = (/ (0.001*i,i = 3,150) /)
            str_s = "r"
        else if (casecode=="c_p") then ! 9
            vt = (/ (0.001*i,i = 10,200) /)
            str_s = "c_P"
        else if (casecode=="delta") then ! 10
            vt = (/ (0.01*i,i = 21,70) /)
            str_s = "delta"
        end if
    
        str_fig=taskcode//'_'//typen//'_'//str_s
    
        open(unit=10,file=str_fig//'_'//str_para//"_ep_d"//".txt")
        open(unit=11,file=str_fig//'_'//str_para//"_ep_c"//".txt")
        open(unit=12,file=str_fig//'_'//str_para//"_theta"//".txt")
        open(unit=13,file=str_fig//'_'//str_para//"_alpha"//".txt")
        open(unit=14,file=str_fig//'_'//str_para//"_exitflag"//".txt")
    
        open(unit=15,file=str_fig//'_'//str_para//"_u"//".txt")
        open(unit=16,file=str_fig//'_'//str_para//"_hs"//".txt")
        open(unit=17,file=str_fig//'_'//str_para//"_hn"//".txt")
        !!!!!!!!!!!!!
        allocate(vepsilon_d(size(vt)))
        allocate(vepsilon_c(size(vt)))
        allocate(vtheta(size(vt)))
        allocate(valpha(size(vt)))
        allocate(vexitflag(size(vt)))
    
        allocate(vu(size(vt)))
        allocate(vhs(size(vt)))
        allocate(vhn(size(vt)))
    
        loop1: do i = 1,size(vt)
            if (casecode=="pstar") then
                parameters(10)=vt(i)
            else if (casecode=="b") then
                parameters(11)=vt(i)
            else if (casecode=="phi") then
                parameters(6)=vt(i)
            else if (casecode=="sigma") then
                parameters(8)=vt(i)
            else if (casecode=="beta") then
                parameters(5)=vt(i)
            else if (casecode=="lambda") then
                parameters(9)=vt(i)
            else if (casecode=="c_f") then
                parameters(12)=vt(i)
            else if (casecode=="r") then
                parameters(3)=vt(i)
            else if (casecode=="c_p") then
                parameters(4)=vt(i)
            else if (casecode=="delta") then
                parameters(7)=vt(i)
            end if
        
            !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
            write(*,"('Parameters are:',15(/f8.3))") parameters
            write(*,"('Parameters are:',15(/f8.3))") P_PARA
            !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
            call Jun01T1(vepsilon_d(i),vepsilon_c(i),vtheta(i),valpha(i),vexitflag(i))
            !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
            call fun_u(vepsilon_d(i),vtheta(i),vu(i))
            call fun_hs(vepsilon_c(i),vepsilon_d(i),vtheta(i),valpha(i),vu(i),vhs(i))
            call fun_hn(vepsilon_c(i),vtheta(i),valpha(i),vu(i),vhn(i))
            !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
            write(10,"(f16.8)") vepsilon_d(i)
            write(11,"(f16.8)") vepsilon_c(i)
            write(12,"(f16.8)") vtheta(i)
            write(13,"(f16.8)") valpha(i)
            write(14,"(L4)") vexitflag(i)
        
            write(15,"(f16.8)") vu(i)
            write(16,"(f16.8)") vhs(i)
            write(17,"(f16.8)") vhn(i)
        
        end do loop1
        !!!!!!!!!!!!!!!!!!!!!!!!!
        deallocate(vepsilon_d)
        deallocate(vepsilon_c)
        deallocate(vtheta)
        deallocate(valpha)
        deallocate(vexitflag)
    
        deallocate(vu)
        deallocate(vhs)
        deallocate(vhn)
        !!!!!!!!!!
        close(UNIT=10,STATUS='KEEP')
        close(UNIT=11,STATUS='KEEP')
        close(UNIT=12,STATUS='KEEP')
        close(UNIT=13,STATUS='KEEP')
        close(UNIT=14,STATUS='KEEP')
    
        close(UNIT=15,STATUS='KEEP')
        close(UNIT=16,STATUS='KEEP')
        close(UNIT=17,STATUS='KEEP')
    end do loop2
    
    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    if (.FALSE.) then 
        ! output to test
        write(*,"('Task code is: ',A10/,'Parameter code is: ',A10/&
            &,'Type is: ',A3/,'epsilon_u = ',F8.2)") taskcode,num_para,typen,epsilon_u
    
        do i = 1,size(parameters)
            write(*,"('Parameters NO.',I5,' is: ',F8.3/)") I,parameters(I)
        end do
    
        do i = 1,size(vt)
            write(*,"('vt NO.',I5,' is: ',F8.3/)") I,vt(I)
        end do
    end if
    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
    pause
end program main_Jun01_Task_1
    
character(len=10) function cell_type(tt)
    implicit none
    integer, intent(in) :: tt
    ! tt is # of type of F(x), which 
    !!!!!!!!!!!!!!!!!!!!!!!!!!!
    if (tt==1) then
        cell_type = "I"
    else if (tt==2) then
        cell_type = "II"
    else if (tt==3) then
        cell_type = "III"
    else if (tt==4) then
        cell_type = "IV"
    else if (tt==5) then
        cell_type = "O"
    end if
    
    return
end
    
character(len=10) function cell_case(cc)
    implicit none
    integer, intent(in) :: cc
    !
    if (cc==1) then
        cell_case = "pstar"
    elseif (cc==2) then
        cell_case = "b"
    else if (cc==3) then
        cell_case = "phi"
    else if (cc==4) then
        cell_case = "sigma"
    else if (cc==5) then
        cell_case = "beta"
    else if (cc==6) then
        cell_case = "lambda"
    else if (cc==7) then 
        cell_case = "c_f"
    else if (cc==8) then
        cell_case = "r"
    else if (cc==9) then
        cell_case = "c_p"
    else if (cc==10) then
        cell_case = "delta"
    end if
    return
end

File: 3 Mathematica/20170614/Others/main_Jun01_Task1.m
% function main_Jun01_Task1(num_para)
% June 1,2017
% May 22-24, 2017
% 20170411 PM
% Mar18,2017
% Task #20170221
% Related to May16/Oct17,2016
% edited in Feb21,2017
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%
close all

taskcode='Jun01T1';
% taskcode='May22T4_1';
num_para='V21';
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
time_start =cputime;
% elasped CPU time when starting the program
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
path0 = pwd;
add_path=strcat(path0,'\Functions\Task_1\');
addpath(add_path);

if strcmp(taskcode,'Jun01T1')
    % mu = 0.08
    % step =0.1
    path_fig = strcat(path0,'\Figures\Task_1\');
    path_data = strcat(path0,'\Data\Task_1\');
    str_para0=strcat('paraJun01',num_para);
elseif strcmp(taskcode,'May22T4_1')
    % mu = 0.012
    % step = 0.15
    path_fig = strcat(path0,'\Figures\Task_4\',...
        'Task_1\');
    path_data = strcat(path0,'\Data\Task_4\',...
        'Task_1\');
    str_para0='paraMay22';
end
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% fsolve algorithm is 'interior-point'
str_algo='LM';
str_para=strcat(str_para0,'-',str_algo);
%% 
cell_typen = {'I';'II';'III';'IV';'O'};% 
cell_case = {'pstar';'b';'phi';'sigma';'beta';'lambda';'c_f';'r';'c_p';'delta'};
% 10 cases

v_tt = 3;% 1:length(cell_typen);
v_cc = 7;% [1;2;3;4;5;6];
for tt = v_tt
    % F(x) type: 'O' 'I' 'II' 'III' 'IV'
    typen=cell_typen{tt};
    for c = v_cc
        close all
        % case code: 'p';'b';'phi';'sigma';'beta';'lambda';'c_f';'r';'c_p';'delta'
        casecode=cell_case{c};
      
        %%
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        % epsilon_u is bar{epsilon}
        % change epsilon_u with different type of function of F(x)
        switch typen
            case 'I'
                epsilon_u=sqrt(3);
            case 'II'
                epsilon_u=sqrt(2)/(sqrt(pi-2));
            case 'III'
                epsilon_u=1;
            case 'IV'
                epsilon_u=(3-sqrt(3))/2;
            case 'O'
                epsilon_u=(pi/4)/(sin(pi/4));
        end
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        if strcmp(taskcode,'Jun01T1')
            % use the parameters of Mar4
            str_func=strcat('myParameter_Jun01',num_para);
            func_handle=str2func(str_func);
            parameters=func_handle(typen,epsilon_u);
            
        elseif strcmp(taskcode,'May22T4_1')
            % use the parameters of May22
            parameters=myParameter_May22T2(typen,epsilon_u);
        end
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
         % parameters=[A;B;r;c_p;beta;phi;delta;sigma;lambda;pstar;b;c_f;mu;B1;B2];
        A=parameters{1};
        B=parameters{2};
        r=parameters{3};
        c_p=parameters{4};
        beta=parameters{5};
        phi=parameters{6};
        delta=parameters{7};
        sigma=parameters{8};
        lambda=parameters{9};
        pstar=parameters{10};
        b=parameters{11};
        c_f=parameters{12};
        mu=parameters{13};
        B1=parameters{14};
        B2=parameters{15};
       %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        % case
        % change parameters with different cases
        % vt is column vector
        switch casecode
            case 'pstar'% 1
                vt=(0.5:0.01:2);
                str_s = 'pstar';
                str_var = '$p^*$';
            case 'b' % 2
                vt=(0.1:0.01:0.5);
                str_s = 'b';
                str_var = '$b$';
            case 'phi' % 3
                vt=(0.05:0.01:0.5);
                str_s = 'phi';
                str_var = '$\phi$';
            case 'sigma' % 4
                vt=(0.01:0.01:0.5);
                str_s = 'sigma';
                str_var = '$\sigma$';
            case 'beta' % 5
                vt=(0.1:0.05:0.7);
                str_s = 'beta';
                str_var = '$\beta$';
            case 'lambda' % 6
                vt=(0.01:0.01:0.5);
                str_s = 'lambda';
                str_var = '$\lambda$';
            case 'c_f' % 7
                % vt=(0.01:0.01:0.5);
                vt=(0.12:0.01:0.5);
                str_s = 'c_F';
                str_var = '$c^F$';
            case 'r' % 8
                vt=(0.005:0.001:0.03);
                str_s = 'r';
                str_var = '$r$';
            case 'c_p' % 9
                vt=(0.01:0.005:0.2);
                str_s = 'c_P';
                str_var = '$c^P$';
            case 'delta' % 10
                vt=(0.1:0.01:0.7);
                str_s = 'delta';
                str_var = '$\delta$';
        end
        str_fig=strcat(taskcode,'_',typen,'_',str_s);
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        vepsilon_d=NaN(size(vt));
        vepsilon_c=NaN(size(vt));
        vtheta=NaN(size(vt));
        valpha=NaN(size(vt));
        vexitflag=NaN(size(vt));
        
        vceq=NaN(4,numel(vt));
        
        vu=NaN(size(vt));
        vhs=NaN(size(vt));
        vhn=NaN(size(vt));
        for i = 1:numel(vt)
            switch casecode
                case 'pstar' % 1
                    pstar=vt(i);
                case 'b' % 2
                    b=vt(i);
                case 'phi' % 3
                    phi=vt(i);
                case 'sigma' % 4
                    sigma=vt(i);
                case 'beta' % 5
                    beta=vt(i);
                case 'lambda' % 6
                    lambda=vt(i);
                case 'c_f' % 7
                    c_f=vt(i);
                case 'r' % 8
                    r=vt(i);
                case 'c_p' % 9
                    c_p=vt(i);
                case 'delta' % 10
                    delta = vt(i);
            end

            %%%%%%%%%%%%%%%%%
            Dis1 = sprintf('We are coputing the case %s \nAnd function F(x) type %s...'...
                ,casecode,typen);
            Dis2 = sprintf('The parameters are:\n');
            Names={'epsilon_u';'A';'B';'r';'c^p';'beta';'phi';'delta';'sigma';'lambda';...
                'pstar';'b';'c^f'};
            Values=[epsilon_u;A;B;r;c_p;beta;phi;delta;sigma;lambda;pstar;b;c_f];
            Tab=table(Values,'RowNames',Names);
            fprintf(Dis1);
            fprintf(Dis2);
            disp(Tab);
            %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
            %%%%%%%%%%% (i)
            % the most important part
            para ={A;B1;B2;c_f;c_p;beta;phi;delta;sigma;lambda;...
                b;r;epsilon_u;mu;pstar;typen;taskcode;...
                str_s;str_para;path_fig;vt;i};
            [vepsilon_d(i),vepsilon_c(i),vtheta(i),valpha(i),vceq(:,i),...
                vexitflag(i)]=Jun01T1(para);
            % solve the equations and get the solution
            

            vu(i) = fun_u(vepsilon_d(i),vtheta(i),lambda,typen,...
                A,B1,B2,epsilon_u);
            vhs(i)=fun_hs(vepsilon_c(i),vepsilon_d(i),vtheta(i),valpha(i),...
                vu(i),lambda,typen,A,B1,B2,epsilon_u);
            vhn(i)=fun_hn(vepsilon_c(i),vtheta(i),valpha(i),vu(i),...
                lambda,typen,A,B1,B2,epsilon_u);
        end
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        % save variables
        save(strcat(path_data,str_fig,'_',str_para,'_ep_d.mat'),'vepsilon_d','-v7.3');
        save(strcat(path_data,str_fig,'_',str_para,'_ep_c_1s.mat'),'vepsilon_c','-v7.3');
        save(strcat(path_data,str_fig,'_',str_para,'_the_1s.mat'),'vtheta','-v7.3');
        save(strcat(path_data,str_fig,'_',str_para,'_alp_1s.mat'),'valpha','-v7.3');
            
        save(strcat(path_data,str_fig,'_',str_para,'_ceq_1s.mat'),'vceq','-v7.3');
        
        save(strcat(path_data,str_fig,'_',str_para,'_u.mat'),'vu','-v7.3');
        save(strcat(path_data,str_fig,'_',str_para,'_hs.mat'),'vhs','-v7.3');
        save(strcat(path_data,str_fig,'_',str_para,'_hn.mat'),'vhn','-v7.3');
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        %% Plot
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        alpha_flag=fun_alphaflag(valpha);
        theta_flag=vtheta>0;
        flag = vexitflag>=0&alpha_flag&theta_flag;
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        figure;
        %
        s1=subplot(2,3,1);
        q_theta=fun_q_theta(vtheta,A,B1,B2);
        tqt=vtheta.*q_theta;
        plot(vt(flag),tqt(flag),'LineWidth',2);
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('$\theta q(\theta)$','interpreter','Latex','FontSize',18,'FontWeight','bold');
        if ~isempty(vt(flag))
            xlim([min(vt(flag)),max(vt(flag))]);
        else
            xlim([min(vt),max(vt)]);
        end
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s2=subplot(2,3,2);
        plot(vt(flag),valpha(flag),'LineWidth',2);
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('$\alpha$','interpreter','Latex','FontSize',18,'FontWeight','bold');
        if ~isempty(vt(flag))
            xlim([min(vt(flag)),max(vt(flag))]);
        else
            xlim([min(vt),max(vt)]);
        end
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s3=subplot(2,3,3);
        vF_epd=fun_F_x(vepsilon_d,typen,epsilon_u);
        vF_epc=fun_F_x(vepsilon_c,typen,epsilon_u);
        hold on 
        plot(vt(flag),vF_epd(flag),'LineWidth',2);
        plot(vt(flag),vF_epc(flag),'LineWidth',2);
        hold off
        leg = legend('$F(\epsilon^d)$','$F(\epsilon^c)$');
        set(leg,'interpreter','Latex');
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('$F(\epsilon)$','interpreter','Latex','FontSize',18,'FontWeight','bold');
        if ~isempty(vt(flag))
            xlim([min(vt(flag)),max(vt(flag))]);
        else
            xlim([min(vt),max(vt)]);
        end
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s4=subplot(2,3,4);
        plot(vt(flag),vu(flag),'LineWidth',2);
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('$u$','interpreter','Latex','FontSize',18,'FontWeight','bold');
        if ~isempty(vt(flag))
            xlim([min(vt(flag)),max(vt(flag))]);
        else
            xlim([min(vt),max(vt)]);
        end
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s5=subplot(2,3,5);
        plot(vt(flag),vhs(flag),'LineWidth',2);
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('$h_s$','interpreter','Latex','FontSize',18,'FontWeight','bold');
        if ~isempty(vt(flag))
            xlim([min(vt(flag)),max(vt(flag))]);
        else
            xlim([min(vt),max(vt)]);
        end
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s6=subplot(2,3,6);
        plot(vt(flag),vhn(flag),'LineWidth',2);
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('$h_n$','interpreter','Latex','FontSize',18,'FontWeight','bold');
        if ~isempty(vt(flag))
            xlim([min(vt(flag)),max(vt(flag))]);
        else
            xlim([min(vt),max(vt)]);
        end
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        grid(s1,'on');grid(s2,'on');grid(s3,'on');
        grid(s4,'on');grid(s5,'on');grid(s6,'on');
        %
        set(gcf,'Position',get(0,'Screensize'));%Maximize figure
        str_fig2 = strcat(path_fig,str_fig,'_',str_para,'_sol_flag');
        saveas(gcf,str_fig2,'tif');
        savefig(gcf,str_fig2);
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        figure;
        %
        s1=subplot(2,3,1);
        q_theta=fun_q_theta(vtheta,A,B1,B2);
        tqt=vtheta.*q_theta;
        plot(vt(vexitflag>=0),tqt(vexitflag>=0),'LineWidth',2);
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('$\theta q(\theta)$','interpreter','Latex',...
            'FontSize',18,'FontWeight','bold');
        if ~isempty(vt(vexitflag>=0))
            xlim([min(vt(vexitflag>=0)),max(vt(vexitflag>=0))]);
        else
            xlim([min(vt),max(vt)]);
        end
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s2=subplot(2,3,2);
        plot(vt(vexitflag>=0),valpha(vexitflag>=0),'LineWidth',2);
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('$\alpha$','interpreter','Latex','FontSize',18,'FontWeight','bold');
        if ~isempty(vt(vexitflag>=0))
            xlim([min(vt(vexitflag>=0)),max(vt(vexitflag>=0))]);
        else
            xlim([min(vt),max(vt)]);
        end
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s3=subplot(2,3,3);
        vF_epd=fun_F_x(vepsilon_d,typen,epsilon_u);
        vF_epc=fun_F_x(vepsilon_c,typen,epsilon_u);
        hold on 
        plot(vt(vexitflag>=0),vF_epd(vexitflag>=0),'LineWidth',2);
        plot(vt(vexitflag>=0),vF_epc(vexitflag>=0),'LineWidth',2);
        hold off
        leg = legend('$F(\epsilon^d)$','$F(\epsilon^c)$');
        set(leg,'interpreter','Latex');
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('$F(\epsilon)$','interpreter','Latex','FontSize',18,'FontWeight','bold');
        if ~isempty(vt(vexitflag>=0))
            xlim([min(vt(vexitflag>=0)),max(vt(vexitflag>=0))]);
        else
            xlim([min(vt),max(vt)]);
        end
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s4=subplot(2,3,4);
        plot(vt(vexitflag>=0),vu(vexitflag>=0),'LineWidth',2);
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('$u$','interpreter','Latex','FontSize',18,'FontWeight','bold');
        if ~isempty(vt(vexitflag>=0))
            xlim([min(vt(vexitflag>=0)),max(vt(vexitflag>=0))]);
        else
            xlim([min(vt),max(vt)]);
        end
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s5=subplot(2,3,5);
        plot(vt(vexitflag>=0),vhs(vexitflag>=0),'LineWidth',2);
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('$h_s$','interpreter','Latex','FontSize',18,'FontWeight','bold');
        if ~isempty(vt(vexitflag>=0))
            xlim([min(vt(vexitflag>=0)),max(vt(vexitflag>=0))]);
        else
            xlim([min(vt),max(vt)]);
        end
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s6=subplot(2,3,6);
        plot(vt(vexitflag>=0),vhn(vexitflag>=0),'LineWidth',2);
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('$h_n$','interpreter','Latex','FontSize',18,'FontWeight','bold');
        if ~isempty(vt(vexitflag>=0))
            xlim([min(vt(vexitflag>=0)),max(vt(vexitflag>=0))]);
        else
            xlim([min(vt),max(vt)]);
        end
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        grid(s1,'on');grid(s2,'on');grid(s3,'on');
        grid(s4,'on');grid(s5,'on');grid(s6,'on');
        %
        set(gcf,'Position',get(0,'Screensize'));%Maximize figure
        str_fig2 = strcat(path_fig,str_fig,'_',str_para,'_solution');
        saveas(gcf,str_fig2,'tif');
        savefig(gcf,str_fig2);
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        % ERRORS
        figure;
        nFn=(1:4);
        surf(vt,nFn,abs(vceq));
        xlabel(str_var,'interpreter','Latex','FontSize',18,'FontWeight','bold');
        ylabel('Fn');zlabel('Error');
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        set(gcf,'Position',get(0,'Screensize'));
        str_fig2 = strcat(path_fig,str_fig,'_',str_para,'_ceq');
        saveas(gcf,str_fig2,'tif');
        savefig(gcf,str_fig2);
        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        % FLAGS
        figure;
        s1 = subplot(2,2,1);
        plot(vt,vexitflag>=0,'LineWidth',2);
        xlabel(str_var,'interpreter','Latex',...
            'FontSize',18,'FontWeight','bold');
        ylabel('$exit flag$','interpreter','Latex',...
            'FontSize',18,'FontWeight','bold');
        xlim([min(vt),max(vt)]);
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s2 = subplot(2,2,2);
        plot(vt,alpha_flag,'LineWidth',2);
        xlabel(str_var,'interpreter','Latex',...
            'FontSize',18,'FontWeight','bold');
        ylabel('$flag:\alpha\subseteq(0,1)$','interpreter',...
            'Latex','FontSize',18,'FontWeight','bold');
        xlim([min(vt),max(vt)]);
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s3 = subplot(2,2,3);
        plot(vt,theta_flag,'LineWidth',2);
        xlabel(str_var,'interpreter','Latex',...
            'FontSize',18,'FontWeight','bold');
        ylabel('$flag:\theta\subseteq(0,\infty)$','interpreter',...
            'Latex','FontSize',18,'FontWeight','bold');
        xlim([min(vt),max(vt)]);
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        s4 = subplot(2,2,4);
        plot(vt,flag,'LineWidth',2);
        xlabel(str_var,'interpreter','Latex',...
            'FontSize',18,'FontWeight','bold');
        ylabel('$flag:\exists solution $','interpreter',...
            'Latex','FontSize',18,'FontWeight','bold');
        xlim([min(vt),max(vt)]);
        str_title = strcat('$',taskcode,',',typen,',',str_para,'$');
        title(str_title,'interpreter','Latex');
        %
        grid(s1,'on');grid(s2,'on');grid(s3,'on');grid(s4,'on');
        %
        set(gcf,'Position',get(0,'Screensize'));%Maximize figure
        str_fig2 = strcat(path_fig,str_fig,'_',str_para,'_exitflag');
        saveas(gcf,str_fig2,'tif');
        savefig(gcf,str_fig2);
        %%
    end
end

close all
time_passby=cputime-time_start;
DCPU = sprintf...
    ('\n Total CPU Time is :%12.3f \n'...
    , time_passby);
disp(DCPU);
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%{
% Send email to notify the fufillment
to = 'huweihugh@gmail.com';
% the efrom address you want to send message to
% you can send to multiple email address, use ; to seperate
subject = strcat('Program_',mfilename,'_done') ;

txt1 = sprintf('%s is done and total cputime is: %12.3f seconds! \n'...
    ,mfilename,time_passby);

txt2=[];
for ii=1:numel(v_tt)
    txt2 = strcat(txt2,'types-',cell_typen(v_tt(ii)),';');
end
txt2r = sprintf('\n The types of function F(x) include:\n %s \n',char(txt2));
txt3=[];
for ii = 1:numel(v_cc)
    txt3 = strcat(txt3,'case-',cell_case(v_cc(ii)),';');
end
txt3r = sprintf('\n The cases of variables include:\n %s \n',char(txt3));

txt4 = sprintf('\n This is automatically sending mail, please DO NOT reply!\n');

message = {txt1;txt2r;txt3r;txt4};

attachment_path = 'D:\GN Luo\20170522\Figures\Task_1\';
filename = 'Apr11_I_b_paraApr11-IP_solution.tif';
attachment = strcat(attachment_path,filename);
% 'folder/attach1.html','attach2.doc'

mysendemail(to,subject,message);
%}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%{
beep 
pause(1)
mydialog
%}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
rmpath(add_path);
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%return